### PR TITLE
Fix popup for reconnecting to the server

### LINKED ISSUE #4272: AutoReConnect
The server is already up! How can I disable this window?

bug

The local server may be temporarily unavailable. Why is there no normal server wait? You need to add another server without accidentally clicking.

I want to see a normal page (not a pop-up!) with the inscription:
```
Waiting for server...

Please return to the server's coverage area or check your connection settings.
```

### PR DESCRIPTION
**Pull request by Roffild**
After receiving an ERROR, the page must be reloaded.

WebView cannot report an error from the server.

Closed #4272

### CODE REVIEW

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 1]
```diff
@@ -470,7 +494,7 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
                                 startActivity(browserIntent)
                                 return true
                             } else {
-                                Log.d(TAG, "No unique cases found to override")
+                                // Log.d(TAG, "No unique cases found to override url=${it}") // Unstoppable flood.
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:497]: Please revert. We don't leave commented code like this, and if you're seeing this statement a lot that is probably an issue with your setup (it is only called for main page navigation/redirects, so with a normal setup you don't see it unless trying to navigate to outside). Some of the log statements you added will generate more lines than this one.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 2]
```diff
@@ -1511,6 +1544,7 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
     }
 
     private fun waitForConnection() {
+        // Called inside an this.loadUrl() [not webView.loadUrl()]
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1547]: This is called from other functions as well? I don't really understand the comment and there seems to be broken Markdown syntax as well.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 3]
```diff
@@ -371,7 +394,8 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
                     errorResponse: WebResourceResponse?
                 ) {
                     Log.e(TAG, "onReceivedHttpError: ${errorResponse?.statusCode} : ${errorResponse?.reasonPhrase} for: ${request?.url}")
-                    if (request?.url.toString() == loadedUrl) {
+                    if (request?.isForMainFrame == true && request.url.toString() == loadedUrl) {
+                        pageError = errorResponse?.statusCode ?: 404
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:398]: Why fallback to 404? HTTP errors can be for various codes; if unknown is a valid response the `pageError` variable should be nullable.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 4]
```diff
@@ -208,6 +209,8 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
     private var mFilePathCallback: ValueCallback<Array<Uri>>? = null
     private var isConnected = false
     private var isShowingError = false
+    private var isPageFinished = false
+    private var pageError = 0 // 0 - OK, <0 - Connection Error, >=400 - HTTP Error
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:213]: What purpose does this have? It is not used in the UI? I can only see it in a log statement when reloading at which point it is no longer relevant because there is reloading 🙃

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 5]
```diff
@@ -903,6 +927,10 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
             checkAndWarnForDisabledLocation()
         }
         changeLog.showChangeLog(this, false)
+
+        if (::loadedUrl.isInitialized) {
+            reloadPage()
+        }
```
Comment (jpelgrom) [inline app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:933]: Please revert. This will cause a reload with state loss anytime the app is resumed, eg. when opening it from the background.

**Review by jpelgrom — CHANGES_REQUESTED** [review]
Demanding merges via Discord is very much not appreciated.

The relation to the issue (which in itself is hard to follow) is unclear. I'm seeing some attempts to handle errors in another callback here and trying to review those, but a clear issue description would be appreciated.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 6]
```diff
@@ -338,11 +350,22 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
...
                     Log.e(TAG, "onReceivedError: errorCode: $errorCode url:$failingUrl")
                     if (failingUrl == loadedUrl) {
+                        pageError = errorCode
+                        showError()
+                    }
+                }
+
+                @RequiresApi(23)
+                override fun onReceivedError(view: WebView?, request: WebResourceRequest?, error: WebResourceError?) {
+                    Log.e(TAG, "onReceivedError: errorCode: ${error?.errorCode} ${error?.description} url:${request?.url}")
```
[the remark this answers is not in the document]
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:360]: This is a copy-paste as in line 351. There is no log flooding here.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 1 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:497]: Without a URL, the information content of this log is poor.
This log shows the correct URL, which has no additional action.

For example, 3 logs per second:  https://community.home-assistant.io/t/apps-cant-access-ha/534210/3

I was receiving more than 10 logs per second.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 7]
```diff
@@ -1394,27 +1425,29 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
...
                         "external"
                     }
+                } else {
+                    waitForConnection()
                 }
-                waitForConnection()
             }
             alert.setNeutralButton(commonR.string.wait) { _, _ ->
-                waitForConnection()
+                reloadPage()
```
[the remark this answers is not in the document]
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1450]: `reloadPage` reloads when there is an ERROR + run `waitForConnection`. Only `waitForConnection` does not work on error.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 5 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:933]: `reloadPage` reloads when there is an ERROR + run `waitForConnection`. Only `waitForConnection` does not work on error.

If the page is normal, then there is no reload. This is needed when returning from Settings, etc.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 2 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1547]: This is not Markdown. This is a pointer to line 1263.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt [hunk 8]
```diff
@@ -329,6 +332,15 @@ class WebViewActivity : BaseActivity(), io.homeassistant.companion.android.webvi
             settings.mediaPlaybackRequiresUserGesture = !presenter.isAutoPlayVideoEnabled()
             settings.userAgentString = settings.userAgentString + " ${HomeAssistantApis.USER_AGENT_STRING}"
             webViewClient = object : TLSWebViewClient(keyChainRepository) {
+                override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
+                    // onReceivedHttpError is getting called before onPageStarted.
+                    // https://issuetracker.google.com/issues/210920403
+                    // onPageStarted is always called before onPageFinished.
+                    // For localhost it is possible to catch onReceived* after onPageFinished !
+                    isPageFinished = false
+                    // Don't pageError=0 here!
```
[the remark this answers is not in the document]
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:341]: The page can load without calling `loadUrl`. Javascript for example.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 4 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:213]: Used in `reloadPage`. It is possible to show more information about the error to the user.

**Comment by Roffild** [comment]
Full fix for:
https://github.com/Roffild/roffild.github.io/assets/38705255/2e049aea-bc77-482c-8355-d03e22345464

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpelgrom) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1450]: The wait button is not for errors. In case of an error, you are supposed to fix it and refresh :)

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (jpelgrom) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:360]: The existing code isn't perfect. I think my comment about this being an unnecessary log statement when nothing is done with it still stands.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 1 above)
Comment (jpelgrom) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:497]: Adding an url could be useful but you're commenting it out so in the end adding no value. And as mentioned, we don't leave in commented code like this.

> I was receiving more than 10 logs per second.

Did you manage to find out what caused it? If so, that could be discussed in an issue and/or changed/fixed in a different PR.

**Comment by jpelgrom** [comment]
> Full fix for: https://github.com/Roffild/roffild.github.io/assets/38705255/2e049aea-bc77-482c-8355-d03e22345464

This still doesn't describe the issue and also doesn't match the code changes which are all over the place. Trying hard to find something here but if we cannot agree on an issue and the value of changes, maybe this should be rejected.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 4 above)
Comment (jpelgrom) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:213]: But you're not showing anything. For debugging with http status codes etc. you may as well use [Chrome remote debugging](http://companion.home-assistant.io/docs/integrations/android-webview/#remote-debugging), the app is never going to match that level of detail.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1450]: If the WebView has an ERROR, then WAIT doesn't work.
Details below...

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:360]: Yes, I agree. But this is not my code:
https://github.com/home-assistant/android/blob/2024.5.2/app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt#L339-L342
https://github.com/home-assistant/android/blob/2024.5.2/app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt#L373-L376

Replace?

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 1 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:497]: The reason is that shouldOverrideUrlLoading() is constantly being called on another thread.
shouldOverrideUrlLoading() is used to filter special URLs.
If the URL is normal, then the flood will be in the log.

It's better to delete this line.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 2 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1547]: This is a note:
```kotlin
// BAD:
this.loadUrl()
this.waitForConnection()

// Good:
this.loadUrl()

// Good:
webView.loadUrl()
this.waitForConnection()
```

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 4 above)
Comment (Roffild) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:213]: Currently used for status in WebView.
It is now possible to show the tip.
`pageError < 0`:
> Check your network connection.

`pageError >= 400` (Android 6.0):
> Check the server configuration.

**Comment by Roffild** [comment]
`showError()` gives a choice to the user. But this choice doesn't know about the WebView's status.

Open http://hass.example.com in your browser. You will get an error. WebView doesn't autoReload. Will you WAIT endlessly (do nothing)? Or do you use page reload (F5)?

`webView.reload()` always reloads the page.

`this.reloadPage()` waits for the page to load or for an error.

File: app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jpelgrom) [reply app/src/main/java/io/homeassistant/companion/android/webview/WebViewActivity.kt:1450]: Sorry, but this still doesn't address the contradiction you've created by making the wait button refresh. I've submitted an alternative in #4412.

**Comment by jpelgrom** [comment]
Closing in favor of #4412
